nabs/callbacks.py

```diff
@@ -0,0 +1,86 @@
+"""
+Callbacks for subscription to the Bluesky ``RunEngine``
+
+This is the LCLS counterpart to `bluesky.callbacks`
+
+Callbacks in this module expect and operate on the `start`, `event`,
+`stop`, etc documents.
+"""
+
+import logging
+
+from bluesky.callbacks.core import CallbackBase, make_class_safe
+
+logger = logging.getLogger(__name__)
+
+
+@make_class_safe(logger=logger)
+class ELogPoster(CallbackBase):
+    """
+    Callback to post messages to the elog, primarily related to plan setup and
+    results. This callback relies on and must be instantiated after the
+    BestEffortCallback and HutchELog.
+
+    To subscribe:
+
+    .. code-block:: python
+
+        elogc = ELogPoster(bec, elog)
+        elogc_uid = RE.subscribe(elogc)
+
+    To enable posting for a specific run:
+
+    .. code-block:: python
+
+        RE(plan(), post=True)
+
+    Parameters
+    ----------
+    bec : BestEffortCallback
+        Instance of BestEffortCallback subscribed to the current RunEngine
+
+    elog : HutchELog
+        Instance of HutchELog connected to the desired logbook
+    """
+    def __init__(self, bec, elog):
+        self._bec = bec
+        self._elog = elog
+
+        # default to elog setting
+        self._send_post = elog.enable_run_posts
+
+        # TO-DO:
+        # - add granular switches as optional arguments
+        # - add error handling for misspelled keys
+        # - add default lists for start doc keys, elog tags?
+
+    def start(self, doc):
+        """ Post plan information on start document"""
+        if 'post' in doc.keys():
+            # Override default if key exists
+            self._send_post = doc['post']
+        else:
+            self._send_post = self._elog.enable_run_posts
+
+        if self._send_post:
+            run_info = str({k: doc[k] for k in ['plan_name', 'plan_args']})
+            logger.info("Posting run start information to elog")
+            self._elog.post(run_info, tags=['plan_info', 'RE'])
```

Comment:
I think exactly what we send to elog here is up for discussion.   Json isn't exactly human-friendly, and the arguments may not immediately inform the reader as to what the plan was.  I'm not entirely sure how to access the raw IPython input, but that'd certainly be useful.  It might be a hook we add to the ipython session rather than the RunEngine

Reply:
Yeah, it's not actually clear that the behavior I'm doing belongs in nabs. If it did, it would involve an `ipython` argument to the class where we'd send the `ipython` instance and do something similar to https://github.com/pcdshub/hutch-python/blob/master/hutch_python/ipython_log.py

Reply:
I think adding the last input is a great idea for this. I think hutch-python should offer a quick way of getting that last input string easily since it has to reformat it to look like the prompt: https://github.com/pcdshub/hutch-python/blob/7f55e3db1f7864d16fd6007bda7082c2511df025/hutch_python/ipython_log.py#L115